src/hotspot/share/runtime/lockStack.cpp

```diff
@@ -67,6 +77,16 @@ void LockStack::verify(const char* msg) const {
     int top = to_index(_top);
     for (int i = 0; i < top; i++) {
       assert(_base[i] != nullptr, "no zapped before top");
+      if (VM_Version::supports_recursive_lightweight_locking()) {
+        oop o = _base[i];
+        for (;i < top - 1; i++) {
```

Comment:
Nit: space after first `;`

Reply:
Though rather than walk the lockstack twice can't we just change the check below to something like:
```
if (VM_Version::supports_recursive_lightweight_locking() && i != j - 1) {
  assert(_base[i] != _base[j], "entries must be unique: %s", msg);
}
```

Reply:
Not sure I understand. 

The change here is that [for every object A, there exists no other object B such that A == B] is changed to [for every contiguous run of object A, there exists no other object B outside that run such that A == B]. Both of these checks are `O(n^2)` in the worst case, and the second is is `O(n)` in the best case. 

I think you have describe the algorithm you envision. Or maybe you want to change what property we are asserting.

Reply:
I thought the issue was that without recursion support we must never find an A and B such that A == B. But with recursion support we can allow A == B if they are adjacent in the lock-stock. So that is what the check is performing, but apparently in two loops, where I was suggesting it can perhaps be done in a single loop.

Reply:
The recursive locking is more strict. There can only be one consecutive range of the same objects. (e.g. `AABCC` is allowed while `AABAA` is not.)

The outer loop finds (the two for loop constructs which works on the index `i`) iterates over each range of consecutive objects which are the same. (Picks out the object at the range start index bumps the index until the object is different, finding the range end index). 
And the inner loop (the for loop construct which works on the index `j`) asserts that no other entry is the same as the object in the this range. (Only walks higher indexes as the lower entries have already been checked in previous iterations).

Reply:
Got it. Sorry I totally misread the code.